apps/arweave/src/ar_mining_server.erl

```diff
@@ -391,56 +409,38 @@ update_cache_limits(State) ->
 	State;
 update_cache_limits(NumActivePartitions, State) ->
-	%% This allows the cache to store enough chunks for 4 concurrent VDF steps per partition.
-	IdealStepsPerPartition = 4,
-	IdealRangesPerStep = 2,
-	%% The number of the sub-chunks in the recall range for packing difficulty = 0 is
-	%% the largest. Each mining worker will reduce it according to its packing difficulty
-	%% in set_cache_limits.
-	ChunksPerRange = ar_mining_worker:recall_range_sub_chunks(0),
-	IdealCacheLimit = ar_util:ceil_int(
-		IdealStepsPerPartition * IdealRangesPerStep * ChunksPerRange * NumActivePartitions, 100),
-
-	{ok, Config} = application:get_env(arweave, config),
-	OverallCacheLimit = case Config#config.mining_server_chunk_cache_size_limit of
-		undefined ->
-			IdealCacheLimit;
-		N ->
-			N
-	end,
-
-	%% We shard the chunk cache across every active worker. Only workers that mine a partition
-	%% included in the current weave are active.
-	NewCacheLimit = max(1, OverallCacheLimit div NumActivePartitions),
+	{MinimumCacheLimitMiB, OverallCacheLimitMiB, PartitionCacheLimit, VDFQueueLimit,
+		GarbageCollectionFrequency} =
+			calculate_cache_limits(NumActivePartitions,  State#state.packing_difficulty),
 
-	case NewCacheLimit == State#state.chunk_cache_limit of
+	case PartitionCacheLimit == State#state.chunk_cache_limit of
 		true ->
 			State;
 		false ->
-			%% Allow enough compute_h0 tasks to be queued to completely refill the chunk cache.
-			VDFQueueLimit = NewCacheLimit div (2 * ChunksPerRange),
 			maps:foreach(
 				fun(_Partition, Worker) ->
-					ar_mining_worker:set_cache_limits(Worker, NewCacheLimit, VDFQueueLimit)
+					ar_mining_worker:set_cache_limits(
+						Worker, PartitionCacheLimit, VDFQueueLimit)
 				end,
 				State#state.workers
 			),
 
 			ar:console(
-				"~nSetting the mining chunk cache size limit to ~B chunks "
-				"(~B chunks per partition).~n", [OverallCacheLimit, NewCacheLimit]),
+				"~nSetting the mining chunk cache size limit to ~B MiB "
+				"(~B sub-chunks per partition).~n",
+					[OverallCacheLimitMiB, PartitionCacheLimit]),
 			?LOG_INFO([{event, update_mining_cache_limits},
-				{limit, OverallCacheLimit}, {per_partition, NewCacheLimit},
+				{overall_limit, OverallCacheLimitMiB}, {per_partition, PartitionCacheLimit},
 				{vdf_queue_limit, VDFQueueLimit}]),
-			case OverallCacheLimit < IdealCacheLimit of
+			case OverallCacheLimitMiB < MinimumCacheLimitMiB of
 				true ->
-					ar:console("~nChunk cache size limit is below minimum limit of ~p. "
+					ar:console("~nChunk cache size limit is below minimum limit of ~p MiB. "
 						"Mining performance may be impacted.~n"
-						"Consider changing the 'mining_server_chunk_cache_size_limit' option.",
-						[IdealCacheLimit]);
+						"Consider changing the 'mining_cache_size_mb' option.",
+						[MinimumCacheLimitMiB]);
 				false -> ok
 			end,
-			GarbageCollectionFrequency = 4 * VDFQueueLimit * 1000,
+
 			GCRef =
 				case State#state.gc_frequency_ms == undefined of
```

Comment:
The logic is not clear here.
We're checking if the `gc_frequency_ms` is defined in the process state, but not checking if the `gc_process_ref` is set (the actual marker that the timer is set here). So there are a possible scenario where we will have two timers set at the same time.
I would recommend saving the actual timer reference in the state and extract the timer initialization function into a separate function. The example may be found in the [P3 ledger implementation](https://github.com/ArweaveTeam/arweave/pull/631/files#diff-cc13c07b3a4f51899a3b95183c92361b26ece97879ac1df7fde8a5482a809452R431) (`handle_continue(?msg_continue_start_shutdown_timer, ...)` @ `ar_p3_ledger.erl` line 431):
- ensure you cancel the existing timer if the timer exists and still running;
- ensure to put the unique marker into timer message ()this is already done here);

When hitting the timer message handler, just do the related work and ensure to call the extracted timer initialization function once again, that will do all the needed things.

Reply:
Got it - Done! https://github.com/ArweaveTeam/arweave/pull/636/commits/cb46fb0912839e365c9f23784b659c3b93e2ad25